apisix/cli/file.lua

```diff
@@ -54,43 +54,61 @@ local function tab_is_array(t)
 end
 
 
+local function var_sub(val)
+    local err
+    local var_used = false
+    -- we use '${{var}}' because '$var' and '${var}' are taken
+    -- by Nginx
+    local new_val = val:gsub("%$%{%{%s*([%w_]+[%:%=]?.-)%s*%}%}", function(var)
+        local i, j = var:find("%:%=")
+        local default
+        if i and j then
+            default = var:sub(i + 2, #var)
+            default = default:gsub('^%s*(.-)%s*$', '%1')
+            var = var:sub(1, i - 1)
+        end
+
+        local v = getenv(var) or default
+        if v then
+            if not exported_vars then
+                exported_vars = {}
+            end
+
+            exported_vars[var] = v
+            var_used = true
+            return v
+        end
+
+        err = "failed to handle configuration: " ..
+              "can't find environment variable " .. var
+        return ""
+    end)
+    return new_val, var_used, err
+end
+
+
 local function resolve_conf_var(conf)
     for key, val in pairs(conf) do
+        -- substitute environment variables from conf keys
+        if type(key) == "string" then
+            local new_key, _, err = var_sub(key)
+            if err then
+                return nil, err
+            end
+            if new_key ~= key then
+                conf.key = nil
+                conf[new_key] = val
```

Comment:
It's better not to change the key in the original table, it will be iterated again

Reply:
I think it's okay to do so for the sake of readability and simplicity we are not dealing with a huge table here, what do you think?

Reply:
Because to avoid modifying the original table we should create a copy of the original table which will again consume CPU cycles.

Reply:
It's a hidden bug here, no one knows it will be iterated two times